File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/services/loaders/LoadJavaClasses.java
package com.github.ol_loginov.heaplibweb.services.loaders;


import com.github.ol_loginov.heaplibweb.repository.heap.HeapEntity;
import com.github.ol_loginov.heaplibweb.repository.heap.HeapRepositories;
import com.github.ol_loginov.heaplibweb.repository.heap.JavaClassEntity;
import lombok.RequiredArgsConstructor;
import org.netbeans.lib.profiler.heap.Heap;
import org.netbeans.lib.profiler.heap.JavaClass;
import org.springframework.transaction.support.TransactionOperations;

import java.util.Optional;
import java.util.concurrent.atomic.AtomicLong;

@RequiredArgsConstructor
class LoadJavaClasses implements Task {
	private final Heap heap;
	private final HeapEntity heapEntity;
	private final TransactionOperations transactionOperations;
	private final HeapRepositories heapRepositories;

	private final AtomicLong passed = new AtomicLong();
	private volatile long total;

	@Override
	public String getText() {
		return "persist java classes: " + passed + "/" + total;
	}

	@Override
	public void run(Callback callback) {
		var all = heap.getAllClasses();
		total = all.size();
		passed.set(0);

		callback.saveProgress(this, true);
		heap.getAllClasses().forEach(clazz -> transactionOperations.executeWithoutResult(st -> {
			persistJavaClass(heapEntity, clazz);
			passed.incrementAndGet();
			callback.saveProgress(this);
		}));
	}

	private void persistJavaClass(HeapEntity heapEntity, JavaClass clazz) {
		var clazzEntity = new JavaClassEntity(
			heapEntity.getId(), clazz.getJavaClassId(),
			clazz.getName(),
			clazz.getAllInstancesSize(),
			clazz.isArray(),
			clazz.getInstanceSize(),
			clazz.getInstancesCount(),
			Task.notReadyValueOnError(clazz::getRetainedSizeByClass),
			Optional.ofNullable(clazz.getSuperClass()).map(JavaClass::getJavaClassId).orElse(null));
		heapRepositories.persist(clazzEntity);
	}
}

File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/services/loaders/LoadJavaClasses.kt
package com.github.ol_loginov.heaplibweb.services.loaders

import com.github.ol_loginov.heaplibweb.hprof.ClassDump
import com.github.ol_loginov.heaplibweb.hprof.HprofFile
import com.github.ol_loginov.heaplibweb.repository.heap.ClassEntity
import com.github.ol_loginov.heaplibweb.repository.heap.HeapRepositories
import com.github.ol_loginov.heaplibweb.support.pretty
import org.slf4j.LoggerFactory
import org.springframework.transaction.support.TransactionOperations

internal class LoadJavaClasses(
    private val hprof: HprofFile,
    private val transactionOperations: TransactionOperations,
    private val heapRepositories: HeapRepositories,
    private val classDumpLookup: ClassDumpLookup
) : Task {
    companion object {
        private val log = LoggerFactory.getLogger(LoadJavaClasses::class.java)
        private val RE_ARRAY_JVM_PREFIX = Regex("^\\[+[ZCFDBSIJL]")
        private val RE_SLASH = Regex("/")
        fun nullIfZero(long: Long): Long? = if (long == 0L) null else long
    }

    private var passed: Long = 0

    @Volatile
    private var classCount: Int = -1

    override fun getText(): String = "import java classes: ${passed.pretty()}/${classCount.pretty()} classes"

    override fun run(callback: Task.Callback) {
        passed = 0

        callback.saveProgress(this, true)
        val insert = InsertCollector("classes") { list ->
            transactionOperations.executeWithoutResult {
                heapRepositories.classLoader.persistAll(list)
            }
        }

        insert.use {
            val classes = hprof.scanClasses()
            classDumpLookup.putAll(classes)
            classCount = classDumpLookup.classCount

            classes.forEach { loadClass: ClassDump ->
                persistJavaClass(loadClass, insert)
                passed++
                callback.saveProgress(this)
            }
        }

        callback.saveProgress(this, true)
    }

    private fun persistJavaClass(clazz: ClassDump, insert: (ClassEntity) -> Unit) {
        if (log.isDebugEnabled) log.debug("{}", clazz.className.name)

        insert(
            ClassEntity(
                clazz.classObjectId.toLong(),
                clazz.superClassObjectId.toLong(),
                clazz.classLoaderObjectId.toLong(),
                unmangleJvmClassName(clazz.className.orEmpty()),
                null,
                clazz.className.orEmpty().startsWith('['), // clazz.isArray,
                clazz.instanceSize,
                0
            )
        )
    }

    private fun unmangleJvmClassName(className: String): String {
        var out = className
        if (className.startsWith('[')) {
            val arrayMatch = RE_ARRAY_JVM_PREFIX.matchAt(className, 0) ?: return className
            val arrayPrefix = arrayMatch.value

            val tail = className.substring(arrayPrefix.length).trimEnd(';')
            val depth = arrayPrefix.count { it == '[' }
            val type = when (arrayPrefix.last()) {
                'J' -> "long"
                'I' -> "int"
                'S' -> "short"
                'B' -> "byte"
                'D' -> "double"
                'F' -> "float"
                'C' -> "char"
                'Z' -> "boolean"
                'L' -> tail
                else -> throw IllegalArgumentException("$arrayPrefix - invalid array prefix")
            }
            out = type + "[]".repeat(depth)
        }

        return out.replace(RE_SLASH, ".")
    }
}
